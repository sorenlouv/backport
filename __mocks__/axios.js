const axios = jest.genMockFromModule('axios');
axios.get = jest.fn(() => {
  throw new Error('Forgot to mock axios.get');
});
axios.post = jest.fn(() => {
  throw new Error('Forgot to mock axios.post');
});
module.exports = axios;
