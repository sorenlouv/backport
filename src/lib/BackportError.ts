import { Commit } from './sourceCommit/parseSourceCommit';

type ErrorContext =
  | {
      code: 'merge-conflict-exception';
      conflictingFiles: string[];
      commitsWithoutBackports: {
        formatted: string;
        commit: Commit;
      }[];
    }
  | {
      code: 'message-only-exception';
      message: string;
    }
  | {
      code: 'no-branches-exception' | 'abort-conflict-resolution-exception';
    };

function getMessage(errorContext: ErrorContext): string {
  switch (errorContext.code) {
    case 'merge-conflict-exception':
      return `Commit could not be cherrypicked due to conflicts in: ${errorContext.conflictingFiles.join(
        ','
      )}`;
    case 'no-branches-exception':
      return 'There are no branches to backport to. Aborting.';
    case 'abort-conflict-resolution-exception':
      return 'Conflict resolution was aborted by the user';
    case 'message-only-exception':
      return errorContext.message;
  }
}

export class BackportError extends Error {
  errorContext: ErrorContext;
  constructor(errorContextOrString: ErrorContext | string) {
    const errorContext: ErrorContext =
      typeof errorContextOrString === 'string'
        ? { code: 'message-only-exception', message: errorContextOrString }
        : errorContextOrString;
    const message = getMessage(errorContext);
    super(message);
    Error.captureStackTrace(this, BackportError);
    this.name = 'BackportError';
    this.message = message;
    this.errorContext = errorContext;
  }
}
