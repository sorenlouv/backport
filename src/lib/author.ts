import { Commit } from '../entrypoint.module';
import { ValidConfigOptions } from '../options/options';
import { getGitConfig, getLocalSourceRepoPath } from './git';

export type GitConfigAuthor = { name?: string; email?: string };
export async function getGitConfigAuthor(
  options: ValidConfigOptions
): Promise<GitConfigAuthor | undefined> {
  const localRepoPath = await getLocalSourceRepoPath(options);
  if (!localRepoPath) {
    return;
  }

  return {
    name: await getGitConfig({ dir: localRepoPath, key: 'user.name' }),
    email: await getGitConfig({ dir: localRepoPath, key: 'user.email' }),
  };
}

export type CommitAuthor = Required<GitConfigAuthor>;
export function getCommitAuthor({
  options,
  commit,
  gitConfigAuthor,
}: {
  options: ValidConfigOptions;
  commit: Commit;
  gitConfigAuthor?: GitConfigAuthor;
}): CommitAuthor {
  if (options.resetAuthor) {
    return {
      name: options.authenticatedUsername,
      email: `<${options.authenticatedUsername}@users.noreply.github.com>`,
    };
  }

  return {
    name: options.gitAuthorName ?? gitConfigAuthor?.name ?? commit.author.name,
    email:
      options.gitAuthorEmail ?? gitConfigAuthor?.email ?? commit.author.email,
  };
}
