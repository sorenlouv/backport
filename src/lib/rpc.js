const promisify = require('es6-promisify');
const process = require('child_process');
const fs = require('fs');
const mkdirp = require('mkdirp');
const findUp = require('find-up');

module.exports = {
  exec: promisify(process.exec),
  writeFile: promisify(fs.writeFile),
  readFile: promisify(fs.readFile),
  stat: promisify(fs.stat),
  statSync: fs.statSync,
  mkdirp: promisify(mkdirp),
  findUp
};
