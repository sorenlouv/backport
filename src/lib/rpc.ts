import { promisify } from 'util';
import { exec as execLegacy } from 'child_process';
import fs from 'fs';
import mkdirpLegacy from 'mkdirp';

const execAsPromised = promisify(execLegacy);
const mkdirpAsPromised = promisify(mkdirpLegacy);

export function exec(cmd: string, options?: {}) {
  return execAsPromised(cmd, { maxBuffer: 100 * 1024 * 1024, ...options });
}
export const writeFile = promisify(fs.writeFile);
export const readFile = promisify(fs.readFile);
export const stat = promisify(fs.stat);
export const chmod = promisify(fs.chmod);
export const statSync = fs.statSync;
export const mkdirp = (path: string) => mkdirpAsPromised(path);
