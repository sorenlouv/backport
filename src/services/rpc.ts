import { promisify } from 'util';
import { exec as execOriginal } from 'child_process';
import fs from 'fs';
import mkdirpOriginal from 'mkdirp';

const execAsPromised = promisify(execOriginal);
const mkdirpAsPromised = promisify(mkdirpOriginal);

export function exec(cmd: string, options?: {}) {
  return execAsPromised(cmd, { maxBuffer: 100 * 1024 * 1024, ...options });
}
export const writeFile = promisify(fs.writeFile);
export const readFile = promisify(fs.readFile);
export const stat = promisify(fs.stat);
export const chmod = promisify(fs.chmod);
export const mkdirp = (path: string) => mkdirpAsPromised(path);
