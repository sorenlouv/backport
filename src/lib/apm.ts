import apm from 'elastic-apm-node';
//@ts-expect-error
import { NoopTransport } from 'elastic-apm-node/lib/noop-transport';
import { accessTokenReplacer } from './logger';

const environment = process.env.NODE_ENV || 'production';

apm
  .start({
    serviceName: 'backport',
    secretToken: 'GZui3tX4jFYjszDweu',
    serverUrl:
      'https://f27ab01db9584e008c443b7137d16425.apm.europe-west2.gcp.elastic-cloud.com:443',
    logLevel: 'off',
    captureBody: 'all',
    errorOnAbortedRequests: false,
    logUncaughtExceptions: false,
    environment,
  })
  // remove access token
  .addFilter((payload) => {
    return JSON.parse(JSON.stringify(payload, accessTokenReplacer));
  });

export function disableApm() {
  // hack to disable APM telemetry after loaded config
  //@ts-expect-error
  apm._transport = new NoopTransport();
}
