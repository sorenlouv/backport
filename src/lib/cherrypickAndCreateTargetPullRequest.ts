import chalk from 'chalk';
import { isEmpty, difference, flatten } from 'lodash';
import { ValidConfigOptions } from '../options/options';
import { BackportError } from './BackportError';
import { CommitAuthor, getCommitAuthor } from './author';
import { spawnPromise } from './child-process-promisified';
import { getRepoPath } from './env';
import { getCommitsWithoutBackports } from './getCommitsWithoutBackports';
import {
  cherrypick,
  createBackportBranch,
  deleteBackportBranch,
  pushBackportBranch,
  getUnstagedFiles,
  commitChanges,
  getConflictingFiles,
  getRepoForkOwner,
  ConflictingFiles,
  getIsMergeCommit,
  getShasInMergeCommit,
} from './git';
import { getFirstLine, getShortSha } from './github/commitFormatters';
import { addAssigneesToPullRequest } from './github/v3/addAssigneesToPullRequest';
import { addLabelsToPullRequest } from './github/v3/addLabelsToPullRequest';
import { addReviewersToPullRequest } from './github/v3/addReviewersToPullRequest';
import {
  createPullRequest,
  getTitle,
  getPullRequestBody,
  PullRequestPayload,
} from './github/v3/createPullRequest';
import { enablePullRequestAutoMerge } from './github/v4/enablePullRequestAutoMerge';
import { fetchCommitBySha } from './github/v4/fetchCommits/fetchCommitBySha';
import { validateTargetBranch } from './github/v4/validateTargetBranch';
import { consoleLog, logger } from './logger';
import { ora, Ora } from './ora';
import { confirmPrompt } from './prompts';
import { sequentially } from './sequentially';
import { Commit } from './sourceCommit/parseSourceCommit';

function getSourceBranchFromCommits(commits: Commit[]) {
  // sourceBranch should be the same for all commits, so picking `sourceBranch` from the first commit should be fine ðŸ¤ž
  // this is specifically needed when backporting a PR like `backport --pr 123` and the source PR was merged to a non-default (aka non-master) branch.
  const { sourceBranch } = commits[0];
  return sourceBranch;
}

export async function cherrypickAndCreateTargetPullRequest({
  options,
  commits,
  targetBranch,
}: {
  options: ValidConfigOptions;
  commits: Commit[];
  targetBranch: string;
}): Promise<{ url: string; number: number; didUpdate: boolean }> {
  const backportBranch = getBackportBranchName(targetBranch, commits);
  const repoForkOwner = getRepoForkOwner(options);
  consoleLog(`\n${chalk.bold(`Backporting to ${targetBranch}:`)}`);

  await validateTargetBranch({ ...options, branchName: targetBranch });
  await createBackportBranch({
    options,
    sourceBranch: getSourceBranchFromCommits(commits),
    targetBranch,
    backportBranch,
  });

  const commitsFlattened = flatten(
    await Promise.all(commits.map((c) => getMergeCommits(options, c)))
  );

  await sequentially(commitsFlattened, (commit) =>
    waitForCherrypick(options, commit, targetBranch)
  );

  if (options.dryRun) {
    ora(options.interactive).succeed('Dry run complete');
    return { url: 'https://localhost/dry-run', didUpdate: false, number: 1337 };
  }

  await pushBackportBranch({ options, backportBranch });
  await deleteBackportBranch({ options, backportBranch });

  const prPayload: PullRequestPayload = {
    owner: options.repoOwner,
    repo: options.repoName,
    title: getTitle({ options, commits, targetBranch }),
    body: getPullRequestBody({ options, commits, targetBranch }),
    head: `${repoForkOwner}:${backportBranch}`, // eg. sqren:backport/7.x/pr-75007
    base: targetBranch, // eg. 7.x
  };

  const targetPullRequest = await createPullRequest({ options, prPayload });

  // add assignees to target pull request
  const assignees = options.autoAssign
    ? [options.authenticatedUsername]
    : options.assignees;

  if (options.assignees.length > 0) {
    await addAssigneesToPullRequest(
      options,
      targetPullRequest.number,
      assignees
    );
  }

  // add reviewers to target pull request
  if (options.reviewers.length > 0) {
    await addReviewersToPullRequest(
      options,
      targetPullRequest.number,
      options.reviewers
    );
  }

  // add labels to target pull request
  if (options.targetPRLabels.length > 0) {
    await addLabelsToPullRequest(
      options,
      targetPullRequest.number,
      options.targetPRLabels
    );
  }

  // make PR auto mergable
  if (options.autoMerge) {
    await enablePullRequestAutoMerge(options, targetPullRequest.number);
  }

  // add labels to source pull requests
  if (options.sourcePRLabels.length > 0) {
    const promises = commits.map((commit) => {
      if (commit.sourcePullRequest) {
        return addLabelsToPullRequest(
          options,
          commit.sourcePullRequest.number,
          options.sourcePRLabels
        );
      }
    });

    await Promise.all(promises);
  }

  consoleLog(`View pull request: ${targetPullRequest.url}`);

  return targetPullRequest;
}

async function getMergeCommits(
  options: ValidConfigOptions,
  commit: Commit
): Promise<Commit[]> {
  const { sha } = commit.sourceCommit;
  if (!options.mainline) {
    const isMergeCommit = await getIsMergeCommit(options, sha);
    if (isMergeCommit) {
      const shas = await getShasInMergeCommit(options, sha);
      return Promise.all(
        shas.reverse().map((sha) => fetchCommitBySha({ ...options, sha }))
      );
    }
  }

  return [commit];
}

/*
 * Returns the name of the backport branch without remote name
 *
 * Examples:
 * For a single PR: `backport/7.x/pr-1234`
 * For a single commit: `backport/7.x/commit-abcdef`
 * For multiple: `backport/7.x/pr-1234_commit-abcdef`
 */
function getBackportBranchName(targetBranch: string, commits: Commit[]) {
  const refValues = commits
    .map((commit) =>
      commit.sourcePullRequest
        ? `pr-${commit.sourcePullRequest.number}`
        : `commit-${getShortSha(commit.sourceCommit.sha)}`
    )
    .join('_')
    .slice(0, 200);
  return `backport/${targetBranch}/${refValues}`;
}

async function waitForCherrypick(
  options: ValidConfigOptions,
  commit: Commit,
  targetBranch: string
) {
  const spinnerText = `Cherry-picking: ${chalk.greenBright(
    getFirstLine(commit.sourceCommit.message)
  )}`;
  const cherrypickSpinner = ora(options.interactive, spinnerText).start();
  const commitAuthor = getCommitAuthor({ options, commit });

  await cherrypickAndHandleConflicts({
    options,
    commit,
    commitAuthor,
    targetBranch,
    cherrypickSpinner,
  });

  // Conflicts should be resolved and files staged at this point

  try {
    // Run `git commit` in case conflicts were not manually committed
    await commitChanges({ options, commit, commitAuthor });
    cherrypickSpinner.succeed();
  } catch (e) {
    cherrypickSpinner.fail();
    throw e;
  }
}

async function cherrypickAndHandleConflicts({
  options,
  commit,
  commitAuthor,
  targetBranch,
  cherrypickSpinner,
}: {
  options: ValidConfigOptions;
  commit: Commit;
  commitAuthor: CommitAuthor;
  targetBranch: string;
  cherrypickSpinner: Ora;
}) {
  const mergedTargetPullRequest = commit.pullRequestStates.find(
    (pr) => pr.state === 'MERGED' && pr.branch === targetBranch
  );

  let conflictingFiles: ConflictingFiles;
  let unstagedFiles: string[];
  let needsResolving: boolean;

  try {
    ({ conflictingFiles, unstagedFiles, needsResolving } = await cherrypick({
      options,
      sha: commit.sourceCommit.sha,
      mergedTargetPullRequest,
      commitAuthor,
    }));

    // no conflicts encountered
    if (!needsResolving) {
      return;
    }
    // cherrypick failed due to conflicts
    cherrypickSpinner.fail();
  } catch (e) {
    cherrypickSpinner.fail();
    throw e;
  }

  const repoPath = getRepoPath(options);

  // resolve conflicts automatically
  if (options.autoFixConflicts) {
    const autoResolveSpinner = ora(
      options.interactive,
      'Attempting to resolve conflicts automatically'
    ).start();

    const didAutoFix = await options.autoFixConflicts({
      files: conflictingFiles.map((f) => f.absolute),
      directory: repoPath,
      logger,
      targetBranch,
    });

    // conflicts were automatically resolved
    if (didAutoFix) {
      autoResolveSpinner.succeed();
      return;
    }
    autoResolveSpinner.fail();
  }

  const conflictingFilesRelative = conflictingFiles
    .map((f) => f.relative)
    .slice(0, 50);

  const commitsWithoutBackports = await getCommitsWithoutBackports({
    options,
    commit,
    targetBranch,
    conflictingFiles: conflictingFilesRelative,
  });

  if (!options.interactive) {
    throw new BackportError({
      code: 'merge-conflict-exception',
      commitsWithoutBackports,
      conflictingFiles: conflictingFilesRelative,
    });
  }

  consoleLog(
    chalk.bold('\nThe commit could not be backported due to conflicts\n')
  );
  consoleLog(`Please fix the conflicts in ${repoPath}`);

  if (commitsWithoutBackports.length > 0) {
    consoleLog(
      chalk.italic(
        `Hint: Before fixing the conflicts manually you should consider backporting the following pull requests to "${targetBranch}":`
      )
    );

    consoleLog(
      `${commitsWithoutBackports.map((c) => c.formatted).join('\n')}\n\n`
    );
  }

  /*
   * Commit could not be cleanly cherrypicked: Initiating conflict resolution
   */

  if (options.editor) {
    await spawnPromise(options.editor, [repoPath], options.cwd);
  }

  // list files with conflict markers + unstaged files and require user to resolve them
  await listConflictingAndUnstagedFiles({
    retries: 0,
    options,
    conflictingFiles: conflictingFiles.map((f) => f.absolute),
    unstagedFiles,
  });
}

async function listConflictingAndUnstagedFiles({
  retries,
  options,
  conflictingFiles,
  unstagedFiles,
}: {
  retries: number;
  options: ValidConfigOptions;
  conflictingFiles: string[];
  unstagedFiles: string[];
}): Promise<void> {
  const hasUnstagedFiles = !isEmpty(
    difference(unstagedFiles, conflictingFiles)
  );
  const hasConflictingFiles = !isEmpty(conflictingFiles);

  if (!hasConflictingFiles && !hasUnstagedFiles) {
    return;
  }

  // add divider between prompts
  if (retries > 0) {
    consoleLog('\n----------------------------------------\n');
  }

  const header = chalk.reset(`Fix the following conflicts manually:`);

  // show conflict section if there are conflicting files
  const conflictSection = hasConflictingFiles
    ? `Conflicting files:\n${chalk.reset(
        conflictingFiles.map((file) => ` - ${file}`).join('\n')
      )}`
    : '';

  const unstagedSection = hasUnstagedFiles
    ? `Unstaged files:\n${chalk.reset(
        unstagedFiles.map((file) => ` - ${file}`).join('\n')
      )}`
    : '';

  const didConfirm = await confirmPrompt(
    `${header}\n\n${conflictSection}\n${unstagedSection}\n\nPress ENTER when the conflicts are resolved and files are staged`
  );

  if (!didConfirm) {
    throw new BackportError({ code: 'abort-conflict-resolution-exception' });
  }

  const MAX_RETRIES = 100;
  if (retries++ > MAX_RETRIES) {
    throw new Error(`Maximum number of retries (${MAX_RETRIES}) exceeded`);
  }

  const [_conflictingFiles, _unstagedFiles] = await Promise.all([
    getConflictingFiles(options),
    getUnstagedFiles(options),
  ]);

  await listConflictingAndUnstagedFiles({
    retries,
    options,
    conflictingFiles: _conflictingFiles.map((file) => file.absolute),
    unstagedFiles: _unstagedFiles,
  });
}
