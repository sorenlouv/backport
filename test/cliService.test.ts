import rimraf from 'rimraf';
import axios from 'axios';
import nock from 'nock';
import httpAdapter from 'axios/lib/adapters/http';
import * as childProcess from 'child_process';

import * as rpc from '../src/lib/rpc';
import {
  maybeSetupRepo,
  doBackportVersion,
  getCommitBySha,
  getReferenceLong
} from '../src/cli/cliService';
import commitMock from './mocks/commit.json';

axios.defaults.adapter = httpAdapter;

describe('doBackportVersion', () => {
  let addLabelMock: nock.Scope;
  beforeEach(() => {
    addLabelMock = nock('https://api.github.com')
      .post(`/repos/elastic/kibana/issues/1337/labels`, ['backport'])
      .query(true)
      .reply(200, {});
  });

  it('with pull request reference', async () => {
    const createPRMock = nock('https://api.github.com')
      .post(`/repos/elastic/kibana/pulls`, {
        title: '[6.x] myCommitMessage | myOtherCommitMessage',
        body:
          'Backports the following commits to 6.x:\n - myCommitMessage (#1000)\n - myOtherCommitMessage (#2000)',
        head: 'sqren:backport/6.x/pr-1000_pr-2000',
        base: '6.x'
      })
      .query(true)
      .reply(200, {
        number: 1337,
        html_url: 'myHtmlUrl'
      });

    const commits = [
      {
        sha: 'mySha',
        message: 'myCommitMessage',
        pullRequest: 1000
      },
      {
        sha: 'mySha2',
        message: 'myOtherCommitMessage',
        pullRequest: 2000
      }
    ];

    const execSpy = jest.spyOn(childProcess, 'exec');

    const res = await doBackportVersion(
      'elastic',
      'kibana',
      commits,
      '6.x',
      'sqren',
      ['backport']
    );

    expect(res).toMatchSnapshot();
    expect(execSpy.mock.calls).toMatchSnapshot();
    expect(createPRMock.isDone()).toBe(true);
    expect(addLabelMock.isDone()).toBe(true);
  });

  it('without pull request reference', async () => {
    const createPRMock = nock('https://api.github.com')
      .post(`/repos/elastic/kibana/pulls`, {
        title: '[6.x] myCommitMessage',
        body:
          'Backports the following commits to 6.x:\n - myCommitMessage (mySha)',
        head: 'sqren:backport/6.x/commit-mySha',
        base: '6.x'
      })
      .query(true)
      .reply(200, {
        number: 1337,
        html_url: 'myHtmlUrl'
      });

    const commits = [
      {
        sha: 'mySha',
        message: 'myCommitMessage'
      }
    ];

    const execSpy = jest.spyOn(childProcess, 'exec');

    const res = await doBackportVersion(
      'elastic',
      'kibana',
      commits,
      '6.x',
      'sqren'
    );

    expect(res).toMatchSnapshot();
    expect(execSpy.mock.calls).toMatchSnapshot();
    expect(createPRMock.isDone()).toBe(true);
    expect(addLabelMock.isDone()).toBe(false);
  });
});

describe('getCommitBySha', () => {
  it('should return a single commit', async () => {
    nock('https://api.github.com')
      .get(`/search/commits`)
      .query(true)
      .reply(200, {
        items: [commitMock]
      });

    nock('https://api.github.com')
      .get(`/search/issues`)
      .query(true)
      .reply(200, {
        items: []
      });

    const commits = await getCommitBySha('elastic', 'kibana', 'myCommitSha');
    expect(commits).toEqual({
      message: '[Chrome] Bootstrap Angular into document.body (#15158)',
      sha: 'myCommitSha',
      pullRequest: undefined
    });
  });

  it('should throw error if sha does not exist', async () => {
    nock('https://api.github.com')
      .get(`/search/commits`)
      .query(true)
      .reply(200, {
        items: []
      });

    await expect(
      getCommitBySha('elastic', 'kibana', 'myCommitSha')
    ).rejects.toThrowError('No commit found for SHA: myCommitSha');
  });

  it('should add PR number if available', async () => {
    nock('https://api.github.com')
      .get(`/search/commits`)
      .query(true)
      .reply(200, {
        items: [commitMock]
      });

    nock('https://api.github.com')
      .get(`/search/issues`)
      .query(true)
      .reply(200, {
        items: [{ number: 1338 }]
      });

    expect(await getCommitBySha('elastic', 'kibana', 'myCommitSha')).toEqual({
      message: '[Chrome] Bootstrap Angular into document.body (#15158)',
      pullRequest: 1338,
      sha: 'myCommitSha'
    });
  });
});

describe('getReferenceLong', () => {
  it('should return a sha', () => {
    expect(
      getReferenceLong({ sha: 'mySha1234567', message: 'myMessage' })
    ).toEqual('mySha12');
  });

  it('should return a pr', () => {
    expect(
      getReferenceLong({
        pullRequest: 1337,
        sha: 'mySha1234567',
        message: 'myMessage'
      })
    ).toEqual('#1337');
  });
});

describe('maybeSetupRepo', () => {
  it('should delete repo if an error occurs', async () => {
    expect.assertions(1);
    jest.spyOn(rpc, 'mkdirp').mockImplementationOnce(() => {
      throw new Error();
    });

    try {
      await maybeSetupRepo('elastic', 'kibana', 'sqren');
    } catch (e) {
      expect(rimraf).toHaveBeenCalledWith(
        '/myHomeDir/.backport/repositories/elastic/kibana',
        expect.any(Function)
      );
    }
  });
});
