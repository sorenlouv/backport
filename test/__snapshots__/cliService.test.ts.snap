// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`doBackportVersion with pull request reference 1`] = `
Object {
  "html_url": "myHtmlUrl",
  "number": 1337,
}
`;

exports[`doBackportVersion with pull request reference 2`] = `
Array [
  Array [
    "git reset --hard && git clean -d --force && git checkout master && git pull origin master",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
  Array [
    "git fetch origin 6.x && git branch backport/6.x/pr-1000_pr-2000 origin/6.x --force && git checkout backport/6.x/pr-1000_pr-2000 ",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
  Array [
    "git cherry-pick mySha",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
  Array [
    "git cherry-pick mySha2",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
  Array [
    "git push sqren backport/6.x/pr-1000_pr-2000:backport/6.x/pr-1000_pr-2000 --force",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
]
`;

exports[`doBackportVersion without pull request reference 1`] = `
Object {
  "html_url": "myHtmlUrl",
  "number": 1337,
}
`;

exports[`doBackportVersion without pull request reference 2`] = `
Array [
  Array [
    "git reset --hard && git clean -d --force && git checkout master && git pull origin master",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
  Array [
    "git fetch origin 6.x && git branch backport/6.x/pr-1000_pr-2000 origin/6.x --force && git checkout backport/6.x/pr-1000_pr-2000 ",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
  Array [
    "git cherry-pick mySha",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
  Array [
    "git cherry-pick mySha2",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
  Array [
    "git push sqren backport/6.x/pr-1000_pr-2000:backport/6.x/pr-1000_pr-2000 --force",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
  Array [
    "git reset --hard && git clean -d --force && git checkout master && git pull origin master",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
  Array [
    "git fetch origin 6.x && git branch backport/6.x/commit-mySha origin/6.x --force && git checkout backport/6.x/commit-mySha ",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
  Array [
    "git cherry-pick mySha",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
  Array [
    "git push sqren backport/6.x/commit-mySha:backport/6.x/commit-mySha --force",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
]
`;
